improved Horus rotary encoder driver and generic behavior

raphaelcoeffic opened the pull request:
Heavily tested on X10 and X7.

fixes #5492

raphaelcoeffic committed:
state table based rotary encoder handler for Horus targets

raphaelcoeffic commented:
To be able test properly the *before* and *after* state, I found useful to use the following traces:
```
diff --git a/radio/src/opentx.cpp b/radio/src/opentx.cpp
index 54a38340a..d85ef72d1 100644
--- a/radio/src/opentx.cpp
+++ b/radio/src/opentx.cpp
@@ -140,6 +140,8 @@ void per10ms()
     if (scrollRE) {
       rePreviousValue = reNewValue;
       putEvent(scrollRE < 0 ? EVT_ROTARY_LEFT : EVT_ROTARY_RIGHT);
+      if (scrollRE < 0) TRACE("-1");
+      else TRACE("+1");
     }
 #if defined(CPUARM)
     // rotary encoder navigation speed (acceleration) detection/calculation
```

raphaelcoeffic commented on radio/src/opentx.cpp:
```diff
@@ -136,7 +136,7 @@ void per10ms()
     static rotenc_t rePreviousValue;
     rotenc_t reNewValue = (ROTARY_ENCODER_NAVIGATION_VALUE / ROTARY_ENCODER_GRANULARITY);
-    int8_t scrollRE = reNewValue - rePreviousValue;
+    rotenc_t scrollRE = reNewValue - rePreviousValue;
```
Please note that this one is probably not necessary (works either way), but I found it cleaner to use the same type here.

bsongis replied:
TODO: check if we use this code on a 8bits platform (we are in an interrupt) and in this case what is rotenc_t type

raphaelcoeffic replied:
There is a define for that, it's basically 8bit an anything but ARM.

raphaelcoeffic commented:
@bsongis I'm testing a way better code right now (using state table), as I could still see some glitches while further testing.

bsongis commented:
no rush, we will keep it for 2.2.2

raphaelcoeffic commented:
exactly, no rush, this time I want it to be as perfect as possible ;-)

raphaelcoeffic committed:
improved generic rotary encoder routine

- ignore direction changes that happen within less than 100ms
- reset velocity on valid direction change

raphaelcoeffic commented:
@bsongis ok, now it's really rock solid. Works 100% for me on X10 and X7.

raphaelcoeffic committed:
removed debug code for rotary encoder

raphaelcoeffic commented:
[X12S_rot_enc_fix.zip](https://github.com/opentx/opentx/files/1564645/X12S_rot_enc_fix.zip)

raphaelcoeffic commented:
[X10_rot_enc_fix.zip](https://github.com/opentx/opentx/files/1564654/X10_rot_enc_fix.zip)

bsongis commented:
We will need more tests from nightly builds, but this seems good
